components/raftstore-v2/src/operation/ready/apply_trace.rs

```diff
@@ -0,0 +1,524 @@
+pub type DataTrace = [u64; DATA_CFS_LEN];
+
+#[derive(Clone, Copy, Default)]
+struct Progress {
+    flushed: u64,
+    /// The index of last entry that has modification to the CF.
+    ///
+    /// If `flushed` == `last_modified`, then all data in the CF is persisted.
+    last_modified: u64,
+}
+
+pub fn cf_offset(cf: &str) -> usize {
+    let cf = if cf.is_empty() { CF_DEFAULT } else { cf };
+    DATA_CFS.iter().position(|c| *c == cf).expect(cf)
+}
+
+#[derive(Default)]
+pub struct ApplyTrace {
+    data_cfs: Box<[Progress; DATA_CFS_LEN]>,
+    admin: Progress,
+    // Index that is issued to be written. It may not be truely persisted.
+    persisted_applied: u64,
+    try_persist: bool,
+}
+
+impl ApplyTrace {
+    fn recover(region_id: u64, engine: &impl RaftEngine) -> Result<(Self, RegionLocalState)> {
+        let mut trace = ApplyTrace::default();
+        // Get all the recorded apply index from data CFs.
+        for (off, cf) in DATA_CFS.iter().enumerate() {
+            // There should be at least one record.
+            let i = engine.get_flushed_index(region_id, cf)?.unwrap();
+            trace.data_cfs[off].flushed = i;
+            trace.data_cfs[off].last_modified = i;
+        }
+        let i = engine.get_flushed_index(region_id, CF_RAFT)?.unwrap();
+        // Index of raft CF means all data before that must be persisted.
+        trace.admin.flushed = i;
+        trace.admin.last_modified = i;
+        trace.persisted_applied = i;
+        let applied_region_state = engine
+            .get_region_state(region_id, trace.admin.flushed)?
+            .unwrap();
+        let data_index = trace.data_index();
+        // If index is not larger than applied_index, it means some CF doesn't have any
+        // data to flush.
+        if trace.admin.flushed < data_index {
+            let region_state = engine.get_region_state(region_id, data_index)?.unwrap();
+            // If there is no admin change, it means `applied_index` is notadvanced in
+            // time. Otherwise, it may be waiting for flush of other tablets in cases like
+            // split.
+            if applied_region_state
+                .get_region()
+                .get_region_epoch()
+                .get_version()
+                == region_state.get_region().get_region_epoch().get_version()
+            {
+                trace.admin.flushed = data_index;
+                trace.admin.last_modified = data_index;
+            }
+        }
+        Ok((trace, applied_region_state))
+    }
+
+    #[inline]
+    fn data_index(&self) -> u64 {
+        self.data_cfs.iter().map(|p| p.flushed).min().unwrap()
+    }
+
+    fn record_flush(&mut self, cf: &str, index: u64) {
+        let off = cf_offset(cf);
+        // Technically it should always be true.
+        if index > self.data_cfs[off].flushed {
+            self.data_cfs[off].flushed = index;
+        }
+    }
+
+    fn record_modify(&mut self, cf: &str, index: u64) {
```

Comment:
It's on purpose to avoid the `_index` suffix, otherwise almost every methods fields in this files will have the suffixes. As long as we all know the file is about indexes, the suffix is not needed.